File: web_version/pmugg_files/edgeSelector.js
// Selects edges and vertices of an object.

ms.edgeSelector = function(driver) {
	this.driver = driver;
};

ms.edgeSelector.prototype.select = function(selection, add, subtract) {
	// This is not generic. Only works with the shape maker.
	var selectedShape = this.driver.selectedShape();
	if (!selectedShape) {
		return;
	}
	var vertices = selectedShape ? selectedShape.getVertices() : [];
	var interiorVertices = [];
	for (var i = 0; i < vertices.length; i++) {
		var v = vertices[i];
		var isInside = selection.isInside(v);
		if (isInside) {
			interiorVertices.push(v);
		}
		ms.selector.updateSelectable(v, add, subtract, isInside);
	}

	// Selecting edges does nothing.
	/* var edges = selectedShape ? selectedShape.getEdges() : [];
	for (var i = 0; i < edges.length; i++) {
		var edgeI = edges[i];
		var intersection = ms.intersector.edgeShapeIntersect(edgeI, selection);
		var vStart = edgeI.getStart().getVertex().getGroup();
		var vEnd = edgeI.getEnd().getVertex().getGroup();
		intersection = intersection ||
				((interiorVertices.indexOf(vStart) != -1) &&
				 (interiorVertices.indexOf(vEnd) != -1));
		ms.selector.updateSelectable(edgeI, add, subtract, intersection);
	} */
};

// Find a selectable that is close to the position.
ms.edgeSelector.findClose = function(selectables, position, scale) {
	for (var i = 0; i < selectables.length; i++) {
		var selectable = selectables[i];
		if (selectable.selectType() == ms.shapeMaker.SelectableTypes.EXAMPLE_SHAPE) {
			var leftEdge = selectable.nearestLeftEdge(position);
			var endpoint = leftEdge.endpoint;
			var face = endpoint && endpoint.getFace();
			var area = face && face.getArea();
			if (area) {
				return face;
			}			
		} else {
			if (selectable.isClose(position, scale)) {
				return selectable;
			}
		}
	}
	return null;
}

// Return vertex near the position. Returns null if no vertices are close.
ms.edgeSelector.prototype.findVertex = function(position, opt_scale) {
	var scale = opt_scale || 1;
	if (!this.driver.selectedShape()) {
		return null;
	}
	return ms.edgeSelector.findClose(this.driver.getVertices(), position, scale);
};

ms.edgeSelector.prototype.pointSelect = function(position, options) {
	var add = options.add || false;
	var subtract = options.subtract || false;
	var scale = options.scale || 1;
	var selectables = this.driver.getSelectables(options);
	var selectable = ms.edgeSelector.findClose(selectables, position, scale);
	if (selectable) {
		if (!selectable.getSelected()) {
			if (!add) {
				this.driver.selectNone(false); // (v.isControl);
			}
			selectable.select();
		} else if (subtract) {
			selectable.deselect();
		}
	} else if (!add && !subtract) {
		this.driver.selectNone();
	}
	return selectable;
};

File: cpp_version/pmugg/edge_selector.cpp
#include "edge_selector.h"
#include "selector.h"
#include "intersector.h"

namespace ms {

EdgeSelector::EdgeSelector(Driver* driver) : driver(driver) {}

void EdgeSelector::select(const Shape& selection, bool add, bool subtract) {
    Shape* selectedShape = driver->selectedShape();
    if (!selectedShape) {
        return;
    }

    auto vertices = selectedShape->getVertices();
    std::vector<Vertex*> interiorVertices;
    
    for (auto* v : vertices) {
        bool isInside = selection.isInside(v);
        if (isInside) {
            interiorVertices.push_back(v);
        }
        Selector::updateSelectable(v, add, subtract, isInside);
    }
}

Selectable* EdgeSelector::findClose(const std::vector<Selectable*>& selectables, 
                                  const Vec2& position, float scale) {
    for (auto* selectable : selectables) {
        if (selectable->selectType() == SelectableTypes::EXAMPLE_SHAPE) {
            auto* leftEdge = selectable->nearestLeftEdge(position);
            auto* endpoint = leftEdge ? leftEdge->endpoint : nullptr;
            auto* face = endpoint ? endpoint->getFace() : nullptr;
            if (face && face->getArea() > 0) {
                return face;
            }
        } else {
            if (selectable->isClose(position, scale)) {
                return selectable;
            }
        }
    }
    return nullptr;
}

Vertex* EdgeSelector::findVertex(const Vec2& position, float scale) {
    if (!driver->selectedShape()) {
        return nullptr;
    }
    return static_cast<Vertex*>(findClose(driver->getVertices(), position, scale));
}

Selectable* EdgeSelector::pointSelect(const Vec2& position, const SelectOptions& options) {
    bool add = options.add;
    bool subtract = options.subtract;
    float scale = options.scale;

    auto selectables = driver->getSelectables(options);
    auto* selectable = findClose(selectables, position, scale);

    if (selectable) {
        if (!selectable->getSelected()) {
            if (!add) {
                driver->selectNone(false);
            }
            selectable->select();
        } else if (subtract) {
            selectable->deselect();
        }
    } else if (!add && !subtract) {
        driver->selectNone();
    }
    
    return selectable;
}

} // namespace ms 